[openvdb] add FindOpenVDB

tmpvar opened the pull request:
This is an attempt at allowing `find_package` to function for openvdb.

However this still breaks when specified as

```
find_package(OpenVDB CONFIG REQUIRED)
```

due to it not being able to find a *config.cmake.

```
CMake Error at D:/work/vcpkg/installed/x64-windows/share/openvdb/vcpkg-cmake-wrapper.cmake:5 (_find_package):
  Could not find a package configuration file provided by "openvdb" with any
  of the following names:

    openvdbConfig.cmake
    openvdb-config.cmake

  Add the installation prefix of "openvdb" to CMAKE_PREFIX_PATH or set
  "openvdb_DIR" to a directory containing one of the above files.  If
  "openvdb" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  D:/work/vcpkg/scripts/buildsystems/vcpkg.cmake:221 (include)
  CMakeLists.txt:4 (find_package)
```

dropping `CONFIG` allows cmake to complete successfully. I took a look around the existing ports but couldn't find one that provided a config - I'd be happy to revise this if someone could provide me with pointers!

tmpvar committed:
[openvdb] add FindOpenVDB

This is an attempt at allowing find_package to function for openvdb.

However this still breaks when specified as

```
find_package(OpenVDB CONFIG REQUIRED)
```

due to it not being able to find a *config.cmake.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/vcpkg?pullRequest=7636) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/vcpkg?pullRequest=7636) before we can accept your contribution.<br/><br/>:x: tmpvar [sign now](https://cla.opensource.microsoft.com/microsoft/vcpkg?pullRequest=7636)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/vcpkg?pullRequest=7636) it.</sub>

tmpvar committed:
[openvdb] add barebones usage file

tmpvar commented:
Oh it just occurred to me that the string that's printed after installing `openvdb`

```
find_package(openvdb CONFIG REQUIRED)
# Note: 1 target(s) were omitted.
target_link_libraries(main PRIVATE Maya CppUnit::cppunit TBB::${COMPONENT} Maya::${COMPONENT})
```
is some sort of auto generated message. I've added a barebones `usage` file, curious to know if this is the correct approach.

tmpvar commented on ports/openvdb/usage:
```diff
@@ -0,0 +1,6 @@
+    find_package(TBB CONFIG REQUIRED)
+    find_package(OPENVDB REQUIRED)
+    target_include_directories(main PRIVATE ${OPENVDB_INCLUDE_DIRS} ${TBB_INCLUDE_DIRS})
+    target_link_libraries(main PRIVATE TBB::tbb ${OpenVDB_LIBRARIES})
```
without the `tbb` related entries here, I get linker errors on VS2017

```
LNK1104	cannot open file 'tbb.lib'
```

cenit replied:
You should find the package tbb inside the openvdb module and “hide” this requirement from the final user

tmpvar replied:
Ah yeah I see how that would be a better experience. Can you point me to a good example port that does this?

cenit commented:
What about the official FindOpenVDB?
https://github.com/AcademySoftwareFoundation/openvdb/blob/master/cmake/FindOpenVDB.cmake

The config file is something that the project should self generate. It’s generally considered not a best practice to write manually a cmake config file.

tmpvar commented:
The reasoning behind rewriting `FindOpenVDB.cmake` is due to how the builds work for openvdb's dependencies. The original find script does not know anything about how the debug/release directories are structured so we're going to end up loading release binaries for debug builds. Due to openvdb depending on `tbb` we've effectively removed the static build option, which limits the some of the utility of the official find script.

There are other issues in the `blosc` port that make it impossible to run a program that has been linked against openvdb in debug.

@cenit I agree that replacing the existing `FindOpenVDB.cmake` script seems less than optimal. So I went back to master and tried to see if using the official `FindOpenVDB.cmake` would be sufficient.

Without also including a `vcpkg-cmake-wrapper.cmake` cmake has no idea where to find this utility.

```
CMake Error at E:/vcpkg/scripts/buildsystems/vcpkg.cmake:266 (_find_package):
  By not providing "FindOpenVDB.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "OpenVDB", but
  CMake did not find one.

  Could not find a package configuration file provided by "OpenVDB" with any
  of the following names:

    OpenVDBConfig.cmake
    openvdb-config.cmake

  Add the installation prefix of "OpenVDB" to CMAKE_PREFIX_PATH or set
  "OpenVDB_DIR" to a directory containing one of the above files.  If
  "OpenVDB" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  CMakeLists.txt:13 (find_package)
```

by adding a `vcpkg-cmake-wrapper.cmake` to openvdb's share directory results in :

```
CMake Error at C:/Program Files/CMake/share/cmake-3.14/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenVDB (missing: OpenVDB_INCLUDE_DIR
  OpenVDB_LIB_COMPONENTS)
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.14/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  E:/vcpkg/installed/x64-windows/share/openvdb/FindOpenVDB.cmake:266 (find_package_handle_standard_args)
  E:/vcpkg/installed/x64-windows/share/openvdb/vcpkg-cmake-wrapper.cmake:5 (_find_package)
  E:/vcpkg/scripts/buildsystems/vcpkg.cmake:221 (include)
  CMakeLists.txt:13 (find_package)
```

Partially resolving this by populating `OPENVDB_INCLUDEDIR` in  `vcpkg-cmake-wrapper.cmake` results in:

```
CMake Error at C:/Program Files/CMake/share/cmake-3.14/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenVDB (missing: OpenVDB_LIB_COMPONENTS) (found version
  "6.1.0")
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.14/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  E:/vcpkg/installed/x64-windows/share/openvdb/FindOpenVDB.cmake:266 (find_package_handle_standard_args)
  E:/vcpkg/installed/x64-windows/share/openvdb/vcpkg-cmake-wrapper.cmake:13 (_find_package)
  E:/vcpkg/scripts/buildsystems/vcpkg.cmake:221 (include)
  CMakeLists.txt:13 (find_package)
```

Which can be somewhat resolved by setting `OPENVDB_LIBRARYDIR` to `installed/<triplet>/lib`, but that means it will always use the release build. I've tried doing the `optimized;...;debug;...` trick that I've seen in other find scripts but that seems to be a vcpkg specific thing, and since the debug path is last it always picks it. 

The `vcpkg-cmake-wrapper.cmake` after all of this looks like:

```cmake
set(OPENVDB_PREV_MODULE_PATH ${CMAKE_MODULE_PATH})
list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR})

find_path(OPENVDB_INCLUDEDIR openvdb/version.h)

get_filename_component(_prefix_path ${OPENVDB_INCLUDEDIR} PATH)

# NOTE: this only supports release builds of openvdb right now
set(OPENVDB_LIBRARYDIR "${_prefix_path}/lib")
unset(_prefix_path)

if(NOT OpenVDB_LIBRARIES OR OPENVDB_LIBRARIES)
  _find_package(${ARGS})
endif()

set(CMAKE_MODULE_PATH ${OPENVDB_PREV_MODULE_PATH})

# be consistent with vcpkg's LIB_INCLUDE_DIRS variable naming
set(OPENVDB_INCLUDE_DIRS ${OPENVDB_INCLUDEDIR})
```

which works with a warning
```
Unable to determine OpenVDB ABI version from OpenVDB installation.  The
  library major version "6" will be inferred.  If this is not correct, use
  add_definitions(-DOPENVDB_ABI_VERSION_NUMBER=N)
```

I'm not really convinced that this is a _better_ way to approach this as it limits the build target to `Release` due to the way deps exist on disk. For example if I run a `Debug` build it fails

![image](https://user-images.githubusercontent.com/46673/62838246-6e815d00-bc2e-11e9-98ab-17e2b6624423.png)

where the contents of the Debug directory:
```
Half-2_3_d.dll
blosc.dll
openvdb.dll
tbb_debug.dll
app.exe
app.ilk
app.pdb
zlibd1.dll
```

and it's using the release build of openvdb

![image](https://user-images.githubusercontent.com/46673/62838265-a7213680-bc2e-11e9-986a-7340ebd9cf05.png)

tmpvar commented:
I guess one way to fix this would be a fairly invasive patch to `FindOpenVDB.cmake`, including the `vcpkg-cmake-wrapper.cmake` from my previous post, applying a patch to fix openexr for debug builds, and sending another PR for `blosc` that handles finding its debug deps properly. 

![image](https://user-images.githubusercontent.com/46673/62838335-77266300-bc2f-11e9-82e8-849e395c6c23.png)

tmpvar committed:
[openvdb] fix debug builds w.r.t OpenEXR

This tends to move away from IlmBase as it is a noop package in
vcpkg and leans on OpenEXR to provide the libs that we require.

tmpvar commented:
I've pushed some changes that fix the `Half-2_3.DLL` issue in the above image (e.g., it will now use `Half-2_3_D.DLL` in openvdb)

cenit commented:
Sorry I don’t have a computer with me right now to help you more (typing from a phone). I can feel your frustration, anyway.
Have you ever used the `*_LIBRARY_DEBUG` and the `*_LIBRARY_RELEASE` symbols? They are optimal to populate the `*_LIBRARY` symbol using the `select_library_configuration` macro, without using explicitly the `debug`/`optimized` keyword... but it might be not enough for you. Just a hint for now, sorry, without any proper way to help you

tmpvar commented:
@cenit yeah, that's effectively what I tried to accomplish with the rewrite of `FindOpenVDB.cmake` (which works, at least for my use case!).

I think once #7649 lands, the 0004 patch file can be updated to simply remove the `FindBlosc.cmake` file.

tmpvar commented:
hrm, maybe I was confused about how ports should be depended on. My assumption was that a find script (e.g., `FindBlosc.cmake`) located in the port's installed dir (e.g., vcpkg/installed/<triplet>/share/blosc/FindBlosc.cmake`) would be somehow in the cmake module search path during configuration of openvdb.

Long story short, I removed the openvdb `FindBlosc.cmake` and that resulted in a configuration failure:

```
CMake Error at E:/vcpkg/scripts/buildsystems/vcpkg.cmake:256 (_find_package):
  By not providing "FindBlosc.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Blosc", but
  CMake did not find one.

  Could not find a package configuration file provided by "Blosc" (requested
  version 1.5) with any of the following names:

    BloscConfig.cmake
    blosc-config.cmake

  Add the installation prefix of "Blosc" to CMAKE_PREFIX_PATH or set
  "Blosc_DIR" to a directory containing one of the above files.  If "Blosc"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  openvdb/CMakeLists.txt:87 (find_package)
```

Did I miss some sort of documentation and by effect making my life harder by assuming things? 😂

cenit commented:
In this PR you took the proper route, adding also the vcpkg wrapper file: https://github.com/tmpvar/vcpkg/blob/ec52dcb5e29f7ba901b3ed4812e2cc205c684cec/ports/openvdb/vcpkg-cmake-wrapper.cmake

Not sure how #7649 has been accepted without that, since adding the find module was the goal since the beginning... @cbezault

cbezault commented:
I accepted it because I was able to make a debug build of a dependent project with it and that was the stated purpose of the PR. I missed the lack of a `vcpkg-wrapper`.
I'll go add one.

cbezault commented:
On second thought, why are we introducing cmake modules? Shouldn't we be creating cmake config files?

PhoebeHui commented:
@tmpvar, can you address the review suggestions? Is work still being done for this PR?

tmpvar commented:
@PhoebeHui it's not clear to me what the next steps are and I haven't looked at this for a while.
